### PR TITLE
Rotated bboxes transforms

### PR DESCRIPTION
**Pull request by AntoineSimoulin**
## Add Transforms support for Rotated Boxes

This PR implements the last transforms for rotated boxes and follows what has been implemented in #9095 and #9084. This PR implements in particular the following modifications :
* Add support for [perspective](https://github.com/pytorch/vision/blob/fcca6ff8bde4a2636102a11b2f7c12e84e005da6/torchvision/transforms/v2/functional/_geometry.py#L1742) for rotated boxes;
* Fix missing tests for [affine](https://github.com/pytorch/vision/blob/main/torchvision/transforms/v2/functional/_geometry.py#L809) transformation and rotated boxes;
* Fix the [`_affine_bounding_boxes_with_expand`](https://github.com/pytorch/vision/blob/fcca6ff8bde4a2636102a11b2f7c12e84e005da6/torchvision/transforms/v2/functional/_geometry.py#L904) function for rotated boxes when `expand=True`;
* Fix [`clamp_bounding_boxes`](https://github.com/pytorch/vision/blob/fcca6ff8bde4a2636102a11b2f7c12e84e005da6/torchvision/transforms/v2/functional/_meta.py#L337) function with behavior detailed below;
* Add support for [elastic](https://github.com/pytorch/vision/blob/fcca6ff8bde4a2636102a11b2f7c12e84e005da6/torchvision/transforms/v2/functional/_geometry.py) for rotated boxes;
* Add support for [crop](https://github.com/pytorch/vision/blob/main/torchvision/transforms/v2/functional/_geometry.py#L1455) for rotated boxes.
* Add missing tests for `TestConvertBoundingBoxFormat`;
* Remove the `SUPPORTED_BOX_FORMATS` and `NEW_BOX_FORMATS` variable in the tests as tests for transform now full cover rotated boxes
* Add support for [sanitize](https://github.com/pytorch/vision/blob/fcca6ff8bde4a2636102a11b2f7c12e84e005da6/torchvision/transforms/v2/functional/_misc.py#L330C5-L330C28) for rotated boxes

## Details on the clamping function

image
image
image
image
image

For the clamping, we re-order the point of the box such that the point with the lowest value on the x-axis is the point 1 (c.f. `_order_bounding_boxes_points`). Given the position of the 4 vertices with respect to the y-axis (c.f. cases above), we are going to adjust the points (x1, y1), (x2, y2), and (x4, y4) to make sure the point (x1, y1) is on the right side of the y-axis. We loop through the four vertices of the rotated box and apply the same operation. In the end we are guaranteed that the bounding box will be within the canvas size and will be completely included within the area of the original box.

We propose some illustration examples below.
image
image

Please note that depending on the order in which we loop through the vertices, we are not guaranteed the output boxes is the box with the largest area that meet the condition above (we might be too aggressive with the clamping. This can occur if the box is largely out of bounds along multiple axis).

## Test plan

Please run the following tests:
```bash
pytest test/test_transforms_v2.py -vvv -k "TestPerspective and test_kernel_bounding_boxes"
pytest test/test_transforms_v2.py -vvv -k "TestPerspective and test_correctness_perspective_bounding_boxes"

pytest test/test_transforms_v2.py -vvv -k "TestAffine and test_transform_bounding_boxes_correctness"

pytest test/test_transforms_v2.py -vvv -k "TestRotate and test_kernel_bounding_boxes"
pytest test/test_transforms_v2.py -vvv -k "TestRotate and test_functional_bounding_boxes_correctness"
pytest test/test_transforms_v2.py -vvv -k "TestRotate and test_transform_bounding_boxes_correctness"

pytest test/test_transforms_v2.py -vvv -k "TestClampBoundingBoxes and test_kernel"
pytest test/test_transforms_v2.py -vvv -k "TestClampBoundingBoxes and test_functional"

pytest test/test_transforms_v2.py -vvv -k "TestElastic and test_kernel_bounding_boxes"

pytest test/test_transforms_v2.py -vvv -k "TestConvertBoundingBoxFormat and test_kernel"
pytest test/test_transforms_v2.py -vvv -k "TestConvertBoundingBoxFormat and test_kernel_noop"
```

## Questions

Please note that this PR introduces a change in the API signature of the [`resize_bounding_boxes`](https://github.com/pytorch/vision/blob/main/torchvision/transforms/v2/functional/_geometry.py#L384) function to add `format: tv_tensors.BoundingBoxFormat` and specialize the operation given whether the box is rotated or not. Contrary to other transforms, this function did typically not received the `format` parameter as input before. Is it possible to add it here? Should we maybe add a default value in this case to preserve the behavior for non-rotated box by default?

## Future work

This PR implements only 4 transforms for rotated boxes. We plan to release the remaining 3 transforms in a subsequent PR.

### CODE REVIEW

File: torchvision/transforms/v2/functional/_meta.py [hunk 1]
```diff
@@ -352,14 +352,122 @@ def _clamp_bounding_boxes(
...
+        x, y = output_xyxyxyxy[..., 0::2], output_xyxyxyxy[..., 1::2]
+        y_max = torch.max(y, dim=1, keepdim=True)[0]
+        _, x1 = (y_max - y).div(y_max).add(x.add(1).mul(100)).min(dim=1)
+        indices = torch.ones_like(output_xyxyxyxy)
+        indices[..., 0] = x1.mul(2)
+        indices.cumsum_(1).remainder_(8)
+    return indices, bounding_boxes.gather(1, indices.to(torch.int64))
+
+
+def area(box: torch.Tensor) -> torch.Tensor:
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_meta.py]: Nit, not strictly needed but helps for consistency with the rest of the helpers

```suggestion
def _area(box: torch.Tensor) -> torch.Tensor:
```

File: torchvision/transforms/v2/functional/_meta.py [hunk 2]
```diff
@@ -352,14 +352,122 @@ def _clamp_bounding_boxes(
     return out_boxes.to(in_dtype)
 
 
+def _order_bounding_boxes_points(
+    bounding_boxes: torch.Tensor, indices: torch.Tensor | None = None
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_meta.py]: I think the `|` operator for annotations will fail on 3.9. Use `Union`, or `from __future__ import annotations`

File: torchvision/transforms/v2/functional/_meta.py [hunk 3]
```diff
@@ -352,14 +352,122 @@ def _clamp_bounding_boxes(
...
+    Returns:
+        tuple[torch.Tensor, torch.Tensor]: A tuple containing:
+            - indices: The indices used for reordering
+            - reordered_boxes: The bounding boxes with reordered points
+    """
+    if indices is None:
+        output_xyxyxyxy = bounding_boxes.clone().reshape(-1, 8)
+        x, y = output_xyxyxyxy[..., 0::2], output_xyxyxyxy[..., 1::2]
+        y_max = torch.max(y, dim=1, keepdim=True)[0]
+        _, x1 = (y_max - y).div(y_max).add(x.add(1).mul(100)).min(dim=1)
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_meta.py]: Nit: we probably don't need to call `clone()` here, since we're never doing any inplace operation?

Also re this line:

```
        _, x1 = (y_max - y).div(y_max).add(x.add(1).mul(100)).min(dim=1)
```

I assume you used the `.div()`, `.add()` etc. methods because we're using them in other ops (for minor perf gains). But here since we're not using the inplace operators, it may be clearer to use normal `+`, `/` operators? Up to you

File: torchvision/transforms/v2/functional/_meta.py [hunk 4]
```diff
@@ -352,14 +352,122 @@ def _clamp_bounding_boxes(
...
+        indices = torch.ones_like(output_xyxyxyxy)
+        indices[..., 0] = x1.mul(2)
+        indices.cumsum_(1).remainder_(8)
+    return indices, bounding_boxes.gather(1, indices.to(torch.int64))
+
+
+def area(box: torch.Tensor) -> torch.Tensor:
+    x1, y1, x2, y2, x3, y3, x4, y4 = box.clone().reshape(-1, 8).unbind(-1)
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_meta.py]: We probably don't need to clone() since reshape is not inplace?

File: torchvision/transforms/v2/functional/_meta.py [hunk 5]
```diff
@@ -352,14 +352,122 @@ def _clamp_bounding_boxes(
...
+    w = (y2 - y1) ** 2 + (x2 - x1) ** 2
+    h = (y3 - y2) ** 2 + (x3 - x2) ** 2
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_meta.py]: Should there be a sqrt() here? I guess it doesn't matter since it's only used for comparison and not for absolute values, but maybe we should add a comment to that effect.

File: torchvision/transforms/v2/functional/_meta.py [hunk 6]
```diff
@@ -352,14 +352,122 @@ def _clamp_bounding_boxes(
...
+    case_d = torch.zeros_like(case_c)
+
+    cond_a = x1.lt(0).logical_and(x2.ge(0)).logical_and(x3.ge(0)).logical_and(x4.ge(0))
+    cond_a = cond_a.logical_and(area(case_a) > area(case_b))
+    cond_a = cond_a.logical_or(x1.lt(0).logical_and(x2.ge(0)).logical_and(x3.ge(0)).logical_and(x4.le(0)))
+    cond_b = x1.lt(0).logical_and(x2.ge(0)).logical_and(x3.ge(0)).logical_and(x4.ge(0))
+    cond_b = cond_b.logical_and(area(case_a) <= area(case_b))
+    cond_b = cond_b.logical_or(x1.lt(0).logical_and(x2.le(0)).logical_and(x3.ge(0)).logical_and(x4.ge(0)))
+    cond_c = x1.lt(0).logical_and(x2.le(0)).logical_and(x3.ge(0)).logical_and(x4.le(0))
+    cond_d = x1.lt(0).logical_and(x2.le(0)).logical_and(x3.le(0)).logical_and(x4.le(0))
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_meta.py]: For all of these, is there a particular reasons to use the methods? If not maybe we can rely on the plain operators like `<` etc.?

File: test/test_transforms_v2.py [hunk 7]
```diff
@@ -647,6 +638,183 @@ def affine_rotated_bounding_boxes(bounding_boxes):
     )
 
 
+def reference_perspective_bounding_boxes(bounding_boxes, *, coefficients, new_canvas_size=None, clamp=True):
```
Comment (NicolasHug) [inline test/test_transforms_v2.py]: Nit: keep this implementation as well as the one for `reference_perspective_rotated_bounding_boxes` within the `TestPerspective` class, since it seems they are only needed there. It might make the diff smaller too!

**Comment by AntoineSimoulin** [comment]
Hey @NicolasHug I publish a fix which should fix the test and address your comments. Here is the list of the modifications:
* Modify the `make_bounding_boxes` function to add clamping and padding, this ensuring that rotated boxes are build within the range of the canvas size;
* Re-placing the `reference_perspective_bounding_boxes` function within the `TestPerspective` class to reduce the number of lines modified in this PR and since this function is only used within the class;
* Decreasing the tightness for the test in `TestAffine` to `atol=1e-5, rtol=2e-5` as the rotation angle had slightly higher variation when computed with the test function. Also let some tolerance for `TestConvertBoundingBoxFormat`;
* Not applying the function `_parallelogram_to_bounding_boxes` to `int` rotated box as the truncation of the point from `float` to `int` does not preserve the rectangular shape of the box;
* Apply clamping after resizing rotated bounding boxes;
* Improve docstring for the `_clamp_rotated_bounding_boxes` function.

Please run the tests with:
```bash
pytest test/test_transforms_v2.py -k box -v
...
2372 passed, 1432 skipped, 5025 deselected in 67.68s (0:01:07)
```

File: torchvision/transforms/v2/functional/_meta.py (hunk 6 above)
Comment (AntoineSimoulin) [reply torchvision/transforms/v2/functional/_meta.py]: I am not sure how to do this as the operation needs to be applied along the axis. Operators such as AND and OR typically reduce the results to a single bolean value. So we will need to use `logical_and`. I can eventually refactor with `<` and `>` if this makes the code more readable.

File: torchvision/transforms/v2/functional/_geometry.py [hunk 8]
```diff
@@ -407,6 +407,10 @@ def _parallelogram_to_bounding_boxes(parallelogram: torch.Tensor) -> torch.Tenso
         torch.int32,
         torch.int64,
     )
+    if int_dtype:
+        # Does not apply the transformation to `int` boxes as the rounding error
+        # will typically not ensure the resulting box has a rectangular shape.
+        return parallelogram.clone()
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_geometry.py:475]: Is it better to return the parallelogram as-is, i.e. *really* not a rectangle, or still try to do the conversion and return something that is closer to a rectangle?

Separately it makes me wonder, maybe we should completely prevent rotated bounding boxes of integer dtype? That would probably make our life a lot easier, and users probably should be running the whole transform pipeline in float anyway, so as to avoid rounding errors compounding?

**Review by NicolasHug — APPROVED** [review]
